cmd/kubeadm/app/util/tokens.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"strings"
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/api"
+)
+
+const (
+	TokenIDLen = 6
+	TokenBytes = 8
+)
+
+func RandBytes(length int) ([]byte, string, error) {
+	b := make([]byte, length)
+	_, err := rand.Read(b)
+	if err != nil {
+		return nil, "", err
+	}
+	// It's only the tokenID that doesn't care about raw byte slice,
+	// so we just encoded it in place and ignore bytes slice where we
+	// do not want it
+	return b, hex.EncodeToString(b), nil
+}
+
+func GenerateToken(s *kubeadmapi.KubeadmConfig) error {
+	_, tokenID, err := RandBytes(TokenIDLen / 2)
+	if err != nil {
+		return err
+	}
+
+	tokenBytes, token, err := RandBytes(TokenBytes)
+	if err != nil {
+		return err
+	}
+
+	s.Secrets.TokenID = tokenID
+	s.Secrets.BearerToken = token
+	s.Secrets.Token = tokenBytes
+	s.Secrets.GivenToken = fmt.Sprintf("%s.%s", tokenID, token)
+	return nil
+}
+
+func UseGivenTokenIfValid(s *kubeadmapi.KubeadmConfig) (bool, error) {
```

Comment:
This makes a lot of sense to me, and the way flags struct is passed around is a very poor design, which is begging for improvement. I believe I've captured all your comments as TODOs now.